indexing-service/src/main/java/org/apache/druid/indexing/common/task/KillUnusedSegmentsTask.java

```diff
@@ -60,6 +61,10 @@ public class KillUnusedSegmentsTask extends AbstractFixedIntervalTask
 {
   private static final Logger LOG = new Logger(KillUnusedSegmentsTask.class);
 
+  // We split this to try and keep each nuke operation relatively short, in the case that either
+  // the database or the storage layer is particularly slow.
+  private static final int SEGMENT_NUKE_BATCH_SIZE = 10_000;
```

Comment:
Should we do some benchmarking to determine the optimal size here? To me, this number seems too large to actually help with the batching.

Unless a lot of segments have been marked as unused, either by a DropRule or the user calling an API, I expect the number of unused segments handled by a `kill` task to typically be under 1000.

So, are the changes in the PR meant to be only upper limit guardrails? Or a perf benefit for normal cases too?

Reply:
Yes, this is based on data from our environment. We currently delete ~2000 rows per second for SQL, so 10000 is ~5 second stall.

By background we currently have a huge backlog maintenance cleanup (100+ million segments), and are arbitrarily limiting kill task Interval size to ~60-80K segments, which takes ~30-40 seconds of stalling the lockbox. But this will take us a _long_ time to perform. Ideally we could issue a kill for X period and the system can go and handle it.

The S3 delete doesn't occupy the TaskLockbox so it can safely be either slow or fast, but, we want the batch size to be large enough that the S3 delete will perform its work in batches (which is 500 segments per batch).

<hr>

I think the right solution here is to provide a default (eg : batch size 500, which will still be vastly more efficient), and then push a `batchSize` parameter into the task definition, to allow for larger (or smaller) batches to be used. I'll add a commit for this.

Reply:
Addressed in 011d04331b203ea63b79ddde7b5423a0370f9c5d, default batch size is now 100 and can input another value if desired during the request, pending deployment on our server to make sure I've implemented it properly.

Reply:
> By background we currently have a huge backlog maintenance cleanup (100+ million segments), and are arbitrarily limiting kill task Interval size to ~60-80K segments, which takes ~30-40 seconds of stalling the lockbox

For such cases, I really feel that deleting by interval (rather than segmentIds) would help significantly. I don't see how the semantics would be any different.

Since you are looking into this, I wonder if you could try evaluating that angle as well. Sorry for going in circles, 😅 . I remember we discussed it but I forget why we decided to not implement it in this PR. Are there any cases that it does not handle?

Also, there is some related work going on in #14662 . We should try to ensure that we do not duplicate any work.

Reply:
I think we decided not to change the semantic of the SQL to be on the safe side. This PR has a small area of impact as it does not touch the SQL query and enforce the batching on a higher level. I saw https://github.com/apache/druid/pull/14662 and I think both PRs are very useful and different.

Reply:
Cool, breaking into batches can definitely not hurt as we are already taking care of not deleting S3 files of undeleted metadata segments. If this helps with making the `TaskLockbox` more granular, we should definitely do this.